### PR TITLE
Refactor to use dropzone native methods upload files

### LINKED ISSUE #20130: Copy and paste the images has list legend
### Feature Description

At the moment, the pasted image in the issue content doesn't appear in the file list below (dropzone).

It's better to make the pasted image added into the file list legend below as the general upload file.

### Screenshots

image

### PR DESCRIPTION
**Pull request by tyroneyeh**
Refer to #20147 to refactor the way to upload attachments and use dropzone to upload natively to achieve a better operating experience
For closes #20130 and #14291 and #14982

### CODE REVIEW

**Comment by wxiaoguang** [comment]
I haven't got time to test it. Some questions:
1. `this.element.parentElement.parentElement....` seems quite fragile. If there is a chance to make the code clear and stable, we should try.
2. Do we still need the prompt `uploading ...` in the textarea when pasting an image from clipboard?
3. `const startPos = editor.selectionStart || editor.getCursor && editor.getCursor('start'), endPos = editor.selectionEnd || editor.getCursor && editor.getCursor('end'), isimage = file.type.startsWith('image/') ? '!' : '', fileName = (isimage ? file.name.replace(/\.[^/.]+$/, '') : file.name);` is too long to read or understand, I do not think it's worth to inline codes aggressively.
4. The different editor's behavior is different, they could be processed separately, instead of using mixed (unclear) `editor.selectionStart || editor.getCursor && editor.getCursor('start')`, in case in the future the editor will be switched to another one.

**Comment by tyroneyeh** [comment]
> I haven't got time to test it. Some questions:
> 
> 1. `this.element.parentElement.parentElement....` seems quite fragile. If there is a chance to make the code clear and stable, we should try.

This section will only be executed if file.editor is not defined, this section is to ensure that file.editor is defined

> 2. Do we still need the prompt `uploading ...` in the textarea when pasting an image from clipboard?

uploading ... do not need to be displayed in the textarea, because the upload progress has been displayed in the dropzone area below, and the image will be added to the textarea after uploading

> 3. `const startPos = editor.selectionStart || editor.getCursor && editor.getCursor('start'), endPos = editor.selectionEnd || editor.getCursor && editor.getCursor('end'), isimage = file.type.startsWith('image/') ? '!' : '', fileName = (isimage ? file.name.replace(/\.[^/.]+$/, '') : file.name);` is too long to read or understand, I do not think it's worth to inline codes aggressively.

Adjusted

> 4. The different editor's behavior is different, they could be processed separately, instead of using mixed (unclear) `editor.selectionStart || editor.getCursor && editor.getCursor('start')`, in case in the future the editor will be switched to another one.

I don't know how to separate, please help

Thanks

File: web_src/js/features/common-global.js [hunk 1]
```diff
@@ -201,13 +190,19 @@ export function initGlobalDropzone() {
           file.uuid = data.uuid;
           const input = $(`<input id="${data.uuid}" name="files" type="hidden">`).val(data.uuid);
           $dropzone.find('.files').append(input);
+          addUploadedFileToEditor(file.editor, file);
         });
         this.on('removedfile', (file) => {
           $(`#${file.uuid}`).remove();
+          if (!file.editor && (file.editor = getAttachedEasyMDE(this.element.parentElement.parentElement.querySelector('textarea')))) {
```
Comment (silverwind) [inline web_src/js/features/common-global.js]: Maybe `this.element.closest('parent').querySelector('textarea')` to make it not so tightly coupled to DOM structure.

File: web_src/js/features/comp/ImagePaste.js [hunk 2]
```diff
@@ -1,94 +1,63 @@
...
-    }
-    endPoint.ch -= oldVal.length;
-    endPoint.ch += newVal.length;
-    editor.setSelection(endPoint, endPoint);
-    editor.focus();
+function clipboardPastedImages(e) {
+  const data = e.clipboardData || e.dataTransfer;
+  if (!data) return [];
+
+  const files = [], datafiles = e.clipboardData && e.clipboardData.items || e.dataTransfer && e.dataTransfer.files;
```
Comment (silverwind) [inline web_src/js/features/comp/ImagePaste.js]:
```suggestion
  const files = [];
  const datafiles = e.clipboardData?.items || e.dataTransfer?.files;
```

File: web_src/js/features/comp/ImagePaste.js [hunk 3]
```diff
@@ -1,94 +1,63 @@
...
+  for (const item of datafiles || []) {
+    const file = (e.clipboardData ? item.getAsFile() : item);
```
Comment (silverwind) [inline web_src/js/features/comp/ImagePaste.js]:
```suggestion
    const file = e.clipboardData ? item.getAsFile() : item;
```

File: web_src/js/features/common-global.js (hunk 1 above)
Comment (tyroneyeh) [reply web_src/js/features/common-global.js]: image

Comment (silverwind) [reply web_src/js/features/common-global.js]: `parent` of course should be replaced with a selector for a common parent node.

Comment (tyroneyeh) [reply web_src/js/features/common-global.js]: Thanks, Fixed on [ad444d3](https://github.com/go-gitea/gitea/pull/20263/commits/ad444d305ddb67065e41474853eb6fb4a225b47d)

**Comment by wxiaoguang** [comment]
About "2. Do we still need the prompt uploading ... in the editor when pasting an image from clipboard? ", I think the answer is still yes. For example, the issue editor's height can be very large (almost full screen), then if a user presses `^V` to paste an image, the dropzone is out of the viewport, they will see no UI feedback.

Before, the `CodeMirrorEditor` and `TextareaEditor` are for such purpose, they are the abstraction layer for different editors to insert/replace placeholders.

Since a PR need 2 approvals to get merged, I am not sure whether there are enough people like this change. If anyone (maybe @silverwind ?) can help to review and get this PR approved in the future, I can help to improve this PR together (in the next few days).

**Comment by tyroneyeh** [comment]
About uploading ... in editor
Origin:
https://user-images.githubusercontent.com/1255041/181421516-a0782b47-ce2e-40ff-a496-a5cb6cb8c181.mp4

New:
https://user-images.githubusercontent.com/1255041/181421462-c4623c89-b811-4aff-bdc4-01dbcd958d40.mp4

**Comment by wxiaoguang** [comment]
I know the uploading will be shown in dropzone. But the problem is, the height of the editor can be very large.

Then you won't see the dropzone in viewport.

image

**Comment by tyroneyeh** [comment]
Add mouse cursor prompt to wait? Is it the solution?
`$('.CodeMirror-lines').css('cursor', 'wait');`

**Comment by wxiaoguang** [comment]
> Add mouse cursor prompt to wait? Is it the solution? `$('.CodeMirror-lines').css('cursor', 'wait');`

Why not just use the placeholder in editor? IMO most editors have placeholders when uploading, including GitHub's editor.

**Comment by tyroneyeh** [comment]
I don't like it very much myself placeholder in editor, and the uploading of pictures is very fast, and the issue of long text is not common.
What do other users think?

**Comment by wxiaoguang** [comment]
But you can not assume every file is small or everyone's network is fast and stable.

Even worst, how about there is a network failure/error? Can users get a feedback from UI?

When writing a general purpose application for end users, all edge cases should be considered.

I am also open for the opinions from other users.

image

**Comment by tyroneyeh** [comment]
The placeholder in the editor will only display uploading... and you can't know the status progress. It's more practical to pull down to see the status, what do you think?

File: web_src/js/features/comp/ImagePaste.js [hunk 4]
```diff
@@ -1,94 +1,72 @@
...
+import {getAttachedEasyMDE} from './EasyMDE.js';
+
+/**
+ *
+ * @param {*} editor
+ * @param {*} file
+ */
+export function addUploadedFileToEditor(editor, file) {
+  if (!editor && file.previewElement) {
+    editor = file.previewElement.closest('form') ? getAttachedEasyMDE(file.previewElement.closest('form').querySelector('textarea')) : getAttachedEasyMDE(file.previewElement.parentElement.parentElement.parentElement.querySelector('textarea'));
```
Comment (silverwind) [inline web_src/js/features/comp/ImagePaste.js]: This part still looks terrible, please find a better solution than `parentElement.parentElement.parentElement`, as I already said, use `el.closest(selector)` to find a common parent node and then `.querySelector` off it.

File: web_src/js/features/comp/ImagePaste.js [hunk 5]
```diff
@@ -1,94 +1,72 @@
...
-      editor.selectionEnd -= oldVal.length;
-      editor.selectionEnd += newVal.length;
+      editor.value += `\n${isimage}[${fileName}](/attachments/${file.uuid})`;
     }
-    editor.selectionStart = editor.selectionEnd;
-    editor.focus();
+  } else if (editor.setSelection) {
+    editor.value(`${editor.value()}\n${isimage}[${fileName}](/attachments/${file.uuid})\n`);
+  } else {
+    editor.value += `${editor.value}\n${isimage}[${fileName}](/attachments/${file.uuid})\n`;
```
Comment (silverwind) [inline web_src/js/features/comp/ImagePaste.js]: Is there a logic behind the final `\n`? I'd generally suggest a helper function to create these strings instead and that can take a boolean whether it should add the newline or not.

File: web_src/js/features/comp/ImagePaste.js [hunk 6]
```diff
@@ -1,94 +1,72 @@
...
-class CodeMirrorEditor {
-  constructor(editor) {
-    this.editor = editor;
+/**
+ * @param editor{EasyMDE}
+ * @param fileUuid
+ */
+export function removeUploadedFileFromEditor(editor, fileUuid) {
+  // the raw regexp is: /!\[[^\]]*]\(\/attachments\/{uuid}\)/
+  const re = new RegExp(`(!|)\\[[^\\]]*]\\(/attachments/${fileUuid}\\)`);
```
Comment (silverwind) [inline web_src/js/features/comp/ImagePaste.js]: Should probably leave a comment on what this regex does.

File: web_src/js/features/comp/ImagePaste.js (hunk 5 above)
Comment (wxiaoguang) [reply web_src/js/features/comp/ImagePaste.js]: The old code has an abstraction layer for different editors (TextareaEditor/CodeMirrorEditor) to do the placeholder inserting/replacing, I think the old code is clearer and more maintainable.

And I still think the placeholder is necessary for users to get obvious feedbacks in all cases.

Comment (tyroneyeh) [reply web_src/js/features/comp/ImagePaste.js]: I refactored again and added placeholders, please review again
Thanks

**Comment by wxiaoguang** [comment]
Thank you. The next steps in my mind:

* A separate PR #20679 to reduce this PR's complexity and make this PR more focused.
* After PR 20679 is merged, I will take a look at the relations between the `editor - dropzone - (paste|drop) events` to see whether it's possible to make the feature more maintainable.
* Write some documents about the editor's test cases, how to fully test the editor's behaviors when the code is changed.

**Comment by wxiaoguang** [comment]
Hmm, sorry for seeing it's closed. If my review caused problems, I apologize, I also worked in my free time, and haven't got time thinking about the editor problems.

There are a lot of historical problems of Gitea's editor module (a lot of bugs and a lot of UI/UX problems), adding more patches may make it more difficult to maintain (and may cause more bugs ...)

If I have time in the future, I may try to refactor it and try to propose some other mechanisms.

Thank you very much.

**Comment by wxiaoguang** [comment]
After "Introduce GitHub markdown editor, keep EasyMDE as fallback #23876" , I'd like to restart the Dropzone related work.

#23876 introduces a general `ComboMarkdownEditor`, I think it would help to make the Dropzone related code simpler in the future, we can avoid many duplicate and fragile code like `initEasyMDEFilePaste`.

Thank you for your work again.

**Comment by silverwind** [comment]
To fix https://github.com/go-gitea/gitea/issues/14982, we should define the textarea as a droppable area so image files can be dragged into. Essentially [this](https://github.com/silverwind/uppie/blob/4978cd24d6fecec740a6aca4842127a3a967031a/index.js#L29-L41) needs to be done (Yes, HTML5 DnD API is weird).

**Pull request closed without merging**